### PR TITLE
[SPARK-32754][SQL][TEST] Unify to `assertEqualJoinPlans` for join reorder suites

### PR DESCRIPTION
**Pull request by wzhfy**
### What changes were proposed in this pull request?

Now three join reorder suites(`JoinReorderSuite`, `StarJoinReorderSuite`, `StarJoinCostBasedReorderSuite`) all contain an `assertEqualPlans` method and the logic is almost the same. We can extract the method to a single place for code simplicity.

### Why are the changes needed?

To reduce code redundancy.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Covered by existing tests.

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala [hunk 1]
```diff
@@ -198,4 +201,23 @@ trait PlanTestBase extends PredicateHelper with SQLHelper { self: Suite =>
         plan1 == plan2
     }
   }
+
+  def assertEqualPlansForJoinReorder(
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala:205]: Can we create a base test suite for join reorder?

Comment (wzhfy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala:205]: yes I also think that's better. Do you think it's necessary to create a separate folder called "joinReorder" under the optimizer package?

Comment (wzhfy) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: @cloud-fan Updated. I add a new folder "joinReorder" under the "optimizer" test package, and put 3 suites and the new base test in it.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala]: Ur, @cloud-fan and @wzhfy , why a new test trait is better than this? The function looks useful in general.

**Review by dongjoon-hyun — COMMENTED** [review]
~The refactoring looks okay. I have only one question on the decision, https://github.com/apache/spark/pull/29594#discussion_r480324425.~

Never mind. The name `assertEqualPlans` was misleading to me, but now I see that it's technically `assertEqualJoinPlan`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/joinReorder/JoinReorderPlanTestBase.scala [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      groundTruthBestPlan: LogicalPlan): Unit = {
+    val analyzed = originalPlan.analyze
+    val optimized = optimizer.execute(analyzed)
+    val expected = EliminateResolvedHint.apply(groundTruthBestPlan.analyze)
+
+    // if this fails, the expected plan itself is incorrect
+    assert(equivalentOutput(analyzed, expected))
+    assert(equivalentOutput(analyzed, optimized))
+
+    compareJoinOrder(optimized, expected)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/joinReorder/JoinReorderPlanTestBase.scala:45]: Oh, got it.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by wzhfy** [comment]
> ~The refactoring looks okay. I have only one question on the decision, [#29594 (comment)](https://github.com/apache/spark/pull/29594#discussion_r480324425).~
> 
> Never mind. The name `assertEqualPlans` was misleading to me, but now I see that it's technically `assertEqualJoinPlan`.

@dongjoon-hyun I think `assertEqualJoinPlan` is a better name, thanks!

**Comment by wzhfy** [comment]
@cloud-fan @dongjoon-hyun I have renamed the method `assertEqualPlans` to `assertEqualJoinPlan`, and moved `compareJoinOrder` and `sameJoinPlan` from `` to `` since they are only used in join reorder test suites.

**Comment by wzhfy** [comment]
@cloud-fan @dongjoon-hyun 
I have renamed the method `assertEqualPlans` to `assertEqualJoinPlan`, and moved `compareJoinOrder` and `sameJoinPlan` from `PlanTest` to `JoinReorderPlanTestBase` since they are only used in join reorder test suites.
Could you please take another look? Thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you so much, @wzhfy and all. Merged to master for Apache Spark 3.1.0 on December 2020.

**Pull request closed without merging**